core/modules/set_metadata.cc

```diff
@@ -28,24 +28,87 @@
 // ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 // POSSIBILITY OF SUCH DAMAGE.
 
+#include <cmath>
+#include <x86intrin.h>
+
 #include "set_metadata.h"
 
 #include "../utils/endian.h"
+#include "../utils/simd.h"
 
 static void CopyFromPacket(bess::PacketBatch *batch, const struct Attr *attr,
                            bess::metadata::mt_offset_t mt_off) {
   int cnt = batch->cnt();
   int size = attr->size;
+  int shift = attr->shift;
+  int pkt_off = attr->offset;
+  int mt_adj = 0;
+  if (shift < 0) {
+    pkt_off += std::abs(shift);
+  } else if (shift > 0) {
+    mt_adj += shift;
+  }
+  size -= std::abs(shift);
 
+  for (int i = 0; i < cnt; i++) {
+    bess::Packet *pkt = batch->pkts()[i];
+    uint8_t *head = pkt->head_data<uint8_t *>(pkt_off);
+    uint8_t *mt_ptr = _ptr_attr_with_offset<uint8_t>(mt_off + mt_adj, pkt);
+    bess::utils::CopySmall(mt_ptr, head, size);
+  }
+}
+
+static void CopyFromPacketWithMask(bess::PacketBatch *batch,
+                                   const struct Attr *attr,
+                                   bess::metadata::mt_offset_t mt_off) {
+  int cnt = batch->cnt();
+  int size = attr->size;
+  mask_t mask = attr->mask;
+  int shift = attr->shift;
   int pkt_off = attr->offset;
+  int mt_adj = 0;
+  if (shift < 0) {
+    pkt_off += std::abs(shift);
+  } else if (shift > 0) {
+    mt_adj += shift;
+  }
 
+  // Copy data
+  size -= std::abs(shift);
   for (int i = 0; i < cnt; i++) {
     bess::Packet *pkt = batch->pkts()[i];
-    char *head = pkt->head_data<char *>();
-    void *mt_ptr;
+    uint8_t *head = pkt->head_data<uint8_t *>(pkt_off);
+    uint8_t *mt_ptr = _ptr_attr_with_offset<uint8_t>(mt_off + mt_adj, pkt);
+    bess::utils::CopySmall(mt_ptr, head, size);
+  }
+  size += std::abs(shift);
 
-    mt_ptr = _ptr_attr_with_offset<value_t>(mt_off, pkt);
-    bess::utils::CopySmall(mt_ptr, head + pkt_off, size);
+  // Apply mask
+  for (int i = 0; i < cnt; i++) {
```

Comment:
Separating the loops had a pretty big (positive) impact on performance iirc (not sure how/why). Will double check and combine if I'm wrong